Sources/NIO/BaseSocketChannel.swift

```diff
@@ -798,7 +798,30 @@ class BaseSocketChannel<T: BaseSocket>: SelectableChannel, ChannelCore {
     // the `reset` event.
     final func reset() {
         self.readEOF()
-        self.close0(error: ChannelError.eof, mode: .all, promise: nil)
+
+        if self.socket.isOpen {
+            assert(self.lifecycleManager.isRegistered)
+            // if the socket is still open, let's try to get the actual socket error from the socket
+            let result: Int32
+            do {
+                result = try self.socket.getOption(level: SOL_SOCKET, name: SO_ERROR)
+                if result != 0 {
+                    // we have a socket error, let's forward
+                    // this path will be executed on Linux (EPOLLERR) & Darwin (ev.fflags != 0)
+                    self.close0(error: IOError(errnoCode: result,
+                                               reason: "connection reset (error set)"),
+                                mode: .all, promise: nil)
+                } else {
+                    // we don't have a socket error, this must be connection reset without an error then
+                    // this path should only be executed on Linux (EPOLLHUP, no EPOLLERR)
+                    self.close0(error: IOError(errnoCode: ECONNRESET,
+                                               reason: "connection reset (no error set)"),
+                                mode: .all, promise: nil)
+                }
+            } catch {
+                self.close0(error: error, mode: .all, promise: nil)
+            }
```

Comment:
This code could be cleaned up, I think, either by stack allocating an existential `Error` that each branch sets, and then calls `close0` with it, or by having the top two branches stack allocate a concrete `IOError`.

Either way we should aim to have fewer of these calls to `close0`.

Reply:
reducing to 2 as we chatted about